### ISSUE TITLE
XLNet: Generalized Autoregressive Pretraining for Language Understanding

### ISSUE THREAD

**Issue by fhiyo** [fhiyo/ml_papers#42]
## 0. 論文

Zhilin Yang ∗1, Zihang Dai ∗12, Yiming Yang 1, Jaime Carbonell 1, Ruslan Salakhutdinov 1, Quoc V. Le 2
1: Carnegie Mellon University, 2: Google Brain

## 1. どんなもの？

language understandingのタスクにおける、BERTに代わる新しいpre-trained modelであるXLNetの提案。
BERTがauto encoding (AE) をpre-trainingの手法に採用しているのに対し、XLNetはauto regressive language modeling (AR) を採用している。

ARの言語モデルとAEの言語モデルのそれぞれの特徴は以下であり、一長一短がある。

- AR

尤度を前向きor後ろ向きに因数分解する形でモデリングする。tokenを順番に読み込ませる形で訓練を行うため、順方向・逆方向の情報を同時に利用できない (一般的に、自然言語処理のタスクは双方向の情報を活用しないと解けない問題が多くある)。
一方、AEと異なり、予測対象の単語同士の依存関係を学習することができる。

- AE

陽に密度推定を行わず、入力の一部を破壊してその入力を再構成するタスクを解くことで言語表現を獲得する。入力文全体を参照しながら再構成を行うため、双方向の情報を活用した文脈表現を獲得できる一方、予測対象の単語同士の依存関係を学習することが難しい問題がある。実際、BERTのpre-trainingは予測対象のtoken同士は (入力文の条件付き) 独立であるという仮定をおいて訓練されている。

そこで、本研究ではARを元にしながらも短所である双方向の情報を学習できない問題を克服するようなタスクを提案している。

### model architecture

Transformer-XLという、SOTAのAR言語モデルを基本としてモデルを構成している。

双方向の情報を獲得するために、元の入力文のpermutationを取ることで単語の位置を入れ替えた文 (ただし、元文の単語の位置はrelative positional encodingを用いて残しておく) を入力とし、その文の後半を予測させるタスクを解くようにする。

入力する文のうち、各単語の隠れ状態はあらかじめ計算しておき (relative positional encodingはpermutationを取る前の位置にのみ依存しており入れ替えの順番には関係ないためあらかじめ計算が行える)、予測を行わない前半部分はmemoryという場所に入れておく。

入力文全体に対するtargetの割合はハイパーパラメータKによって調節する。実験ではK=6で設定している。Kの近似式は以下の通り。
![image](https://user-images.githubusercontent.com/13327262/67637566-14379480-f91f-11e9-8ae5-2f4d8f6e5508.png)

![image](https://user-images.githubusercontent.com/13327262/67635502-76869a00-f90b-11e9-9233-2d8073d56583.png)

pre-trainingの一例。元の入力文が(x\_1, x\_2, x\_3, x\_4)であり、4つの入れ替え順序に対してx\_3を予測する場合を見る。

左上: 入れ替え後の文が(x\_3, x\_2, x\_4, x\_1)である。x\_3よりも左にtokenがないので、memoryからのみ情報を参照してh\_3を計算し、x\_3を予測する。
右上: 入れ替え後の文が(x\_2, x\_4, x\_3, x\_1)である。x\_3よりも左にx\_2, x\_4があるので、memory, x\_2, x\_4を使ってh\_3を計算し、x\_3を予測する。
左下: 入れ替え後の文が(x\_1, x\_4, x\_2, x\_3)である。x\_3よりも左にx\_1, x\_4, x\_2があるので、memory, x\_1, x\_2, x\_4を使ってh\_3を計算し、x\_3を予測する。
右下: 入れ替え後の文が(x\_4, x\_3, x\_1, x\_2)である。x\_3よりも左にx\_4があるので、memory, x\_4を使ってh\_3を計算し、x\_3を予測する。

### two-stream self-attentionについて

このpermutation language modelingは基本的にはtransformerのdecoder側の学習方法を用いるが、permutationにより語順がランダムになっていることにより、ナイーブな方法では本来異なる位置に存在するtokenの分布が全く同じになってしまう。
一つ例を見てみる。異なる2つの順列z\_1, z\_2があり、これらはある時刻tより前は同一で、tで初めて異なる数字になるとする。数式で表すと、

![image](https://user-images.githubusercontent.com/13327262/67636230-f1eb4a00-f911-11e9-8d0f-8557aa6a854c.png)

となる。この2つの順列について時刻tのtokenの生起確率を計算すると、

![image](https://user-images.githubusercontent.com/13327262/67636264-2828c980-f912-11e9-9f6d-0bb879eeaf58.png)

となり、同じ分布が計算されていることが確認できる。
この問題を解決するため、入れ替え後の位置に依存した隠れ状態gを新たに作成し、self-attentionのqueryをgから供給するようなattention (query stream attention) を考える。

![image](https://user-images.githubusercontent.com/13327262/67636768-31686500-f917-11e9-9dc8-c65668cd0154.png)

通常のself-attentionをcontent stream attention (上図のa), queryの元になるcontentsを参照せずにその位置情報だけを参照するように設計したattentionをquery stream attention (上図のb) と呼ぶ。なお、上図の(c)の最下層にあるwはモデルの訓練対象のパラメータベクトルである。この2つのattentionを求め、next-token predictionを以下の計算式から計算することで予測を行う (数式内のgは最上位層のquery stream attention)。

![image](https://user-images.githubusercontent.com/13327262/67636910-b011d200-f918-11e9-9a87-6fbd031e8130.png)

そしてfine-tune時にはquery streamは単純に除去してcontent streamのみを学習すればよい。

### multiple segmentのモデリングについて

XLNetはBERTと異なり、入力する文のsegment encodingはtokenごとにつけるのではなく、tokenの関係ごとにつける。
具体的には、入力文内のペアとなる位置iとjが与えられたとき、iとjが同じsegment内にある場合はs\_{ij} = s\_+という値で、iとjが異なるsegmentにある場合はs\_{ij} = s\_-という値でencodingする。そして (content) attention weightを計算する際には、a\_{ij} = (q\_i + b)^T s\_{ij}の式に従い計算する。各tokenが同じsegmentにあるかどうかのみを興味の対象にすることで、segmentが2つ以上必要なfine-tuneのタスクにも適用できるような設計にしている。

## 2. 先行研究と比べてどこがすごい？

BERTと比較して、right-to-leftな文脈も (そしてその逆のleft-to-rightも同様に) 考慮でき、maskされた単語同士の関係も考慮することができる点。
XLNetで解いているのAR系のタスクは、あるtokenの予測をそれまでに予測した文脈情報を利用して行えるため、例えば"New York is a city"という入力のうち"New York"を予測するタスクの場合、BERTは下図上のように各tokenを独立して尤度を計算するが、XLNetは下図下のようにmaskされた単語同士の依存関係も利用して尤度計算が行え、学習ができる。

![image](https://user-images.githubusercontent.com/13327262/67637255-e270fe80-f91b-11e9-9e72-323221c6c7b3.png)

さらに、他のAR系のpre-trainingの設計はleft-to-rightかright-to-leftかのいずれかしか行えないため、p(York|New)は計算できるが逆方向の依存関係であるp(New|York)は計算できない、という問題があった。XLNetはpermutationを行うことで様々な方向からの依存関係を学習ができる。

先行研究のBERTの短所として、以下の点を述べている。

- pre-training時に[MASK]という人工的なラベルを使用しているため、fine-tune時の実際のデータには出現しないラベルを学習することになる。これがpretrainとfinetuneの不一致を起こす (※が、これがどう影響しているのかを確かめてはいないと思う)。
- maskされた1tokenの予測を行うため、複数の単語の同時確率を学習・推論することができない。つまりBERTは予測したtoken同士は、マスクされていないtokenで条件付けられた元で独立だという仮定を行っている。これは自然言語のtoken同士は高次に、長距離に依存関係があることから考えると過度に単純化された条件である

## 4. どうやって有効だと検証した？

20のタスクで実験を行い、そのうちの18のタスクでSOTAを達成した。

pre-trainingのデータセットとして、BooksCorpus, English Wikipedia, Giga5, ClueWeb2012-B, Common Crawlを用いた。これをsentencepieceで分割することで、トータルで32.89Bのtoken数のpre-training用データセットを獲得した。

一文が長く、難しいタスクであるRACE datasetを用いて先行研究のモデルと比較したところ、大幅なスコアの向上が見られた。これはXLNetが持っているTransformer-XLの構造が通常のTransformerよりも長距離の依存関係を学習できるように設計されたものであるので、その影響を受けたものだろうと思われる。

## 5. 議論はある？

BERTが行っているnext-sentence predictionのタスクをXLNetのpre-trainingに取り入れたところ逆に性能が下がった (RACE datasetを用いたタスクを除く)。

参考: <https://mc.ai/bertを超えたxlnetの紹介/>

## 6. 次に読むべき論文は？

[Transformer-XL: Attentive Language Models Beyond a Fixed-Length Context](https://arxiv.org/abs/1901.02860)
